{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Visualizing ALDEx2 feature differentials using Qurro\n",
    "\n",
    "In this example, we use transcriptomic data from [TCGA](https://portal.gdc.cancer.gov/repository). We downloaded \n",
    "100 gene expression files from lung squamous cell carcinoma (LUSC) and 49 solid tissue normal expression files. We have pre-processed this data into a feature table for ease of use, but the gdc manifest file is also provided for your convenience as well as the script used to aggregate all the files.\n",
    "\n",
    "## Requirements:\n",
    "\n",
    "This notebook requires Qurro, pandas, and biom to be installed for Python. The R packages [ALDEx2](http://bioconductor.org/packages/release/bioc/html/ALDEx2.html) and [dplyr](https://dplyr.tidyverse.org/) are also required for the `run_aldex.R` script.\n",
    "\n",
    "`NOTE`: This breaks with pandas 1.0.0 due to a biom issue\n",
    "\n",
    "`TODO`: *insert TCGA citation here*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 0. Setting up\n",
    "In this section, we replace the output directory with an empty directory. This just lets us run this notebook multiple times, without any tools complaining about overwriting files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clear the output directory so we can write these files there\n",
    "!rm -rf output/*\n",
    "# Since git doesn't keep track of empty directories, create the output/ directory if it doesn't already exist\n",
    "# (if it does already exist, -p ensures that an error won't be thrown)\n",
    "!mkdir -p output"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Filtering the feature table\n",
    "\n",
    "The original feature table has over 60,000 features which is too computationally expensive to process for this example. We will filter this table to use the top 10% of genes by total abundance."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(60483, 149)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-43-5670-11A</th>\n",
       "      <th>TCGA-77-8008-11A</th>\n",
       "      <th>TCGA-18-3410-01A</th>\n",
       "      <th>TCGA-43-6771-11A</th>\n",
       "      <th>TCGA-66-2758-01A</th>\n",
       "      <th>TCGA-90-7767-01A</th>\n",
       "      <th>TCGA-66-2795-01A</th>\n",
       "      <th>TCGA-77-7138-01A</th>\n",
       "      <th>TCGA-39-5019-01A</th>\n",
       "      <th>TCGA-90-6837-11A</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-56-7823-01B</th>\n",
       "      <th>TCGA-77-7142-11A</th>\n",
       "      <th>TCGA-22-5483-11A</th>\n",
       "      <th>TCGA-77-8153-01A</th>\n",
       "      <th>TCGA-85-A4JC-01A</th>\n",
       "      <th>TCGA-39-5040-11A</th>\n",
       "      <th>TCGA-43-6143-11A</th>\n",
       "      <th>TCGA-77-7335-11A</th>\n",
       "      <th>TCGA-85-7698-01A</th>\n",
       "      <th>TCGA-43-2581-01A</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feature-id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003.13</th>\n",
       "      <td>4707</td>\n",
       "      <td>1016</td>\n",
       "      <td>2286</td>\n",
       "      <td>1238</td>\n",
       "      <td>2982</td>\n",
       "      <td>2795</td>\n",
       "      <td>1516</td>\n",
       "      <td>5507</td>\n",
       "      <td>5504</td>\n",
       "      <td>2079</td>\n",
       "      <td>...</td>\n",
       "      <td>1988</td>\n",
       "      <td>1059</td>\n",
       "      <td>1297</td>\n",
       "      <td>1970</td>\n",
       "      <td>2450</td>\n",
       "      <td>1432</td>\n",
       "      <td>1500</td>\n",
       "      <td>1790</td>\n",
       "      <td>3722</td>\n",
       "      <td>2471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005.5</th>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419.11</th>\n",
       "      <td>2019</td>\n",
       "      <td>1731</td>\n",
       "      <td>2123</td>\n",
       "      <td>2121</td>\n",
       "      <td>3868</td>\n",
       "      <td>2722</td>\n",
       "      <td>2130</td>\n",
       "      <td>4834</td>\n",
       "      <td>4557</td>\n",
       "      <td>1264</td>\n",
       "      <td>...</td>\n",
       "      <td>1194</td>\n",
       "      <td>1449</td>\n",
       "      <td>1382</td>\n",
       "      <td>747</td>\n",
       "      <td>1741</td>\n",
       "      <td>3167</td>\n",
       "      <td>1454</td>\n",
       "      <td>1569</td>\n",
       "      <td>2477</td>\n",
       "      <td>1826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457.12</th>\n",
       "      <td>1062</td>\n",
       "      <td>968</td>\n",
       "      <td>1883</td>\n",
       "      <td>655</td>\n",
       "      <td>800</td>\n",
       "      <td>1987</td>\n",
       "      <td>935</td>\n",
       "      <td>898</td>\n",
       "      <td>1014</td>\n",
       "      <td>840</td>\n",
       "      <td>...</td>\n",
       "      <td>449</td>\n",
       "      <td>883</td>\n",
       "      <td>614</td>\n",
       "      <td>678</td>\n",
       "      <td>305</td>\n",
       "      <td>705</td>\n",
       "      <td>917</td>\n",
       "      <td>658</td>\n",
       "      <td>961</td>\n",
       "      <td>470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460.15</th>\n",
       "      <td>204</td>\n",
       "      <td>202</td>\n",
       "      <td>1923</td>\n",
       "      <td>205</td>\n",
       "      <td>706</td>\n",
       "      <td>2191</td>\n",
       "      <td>956</td>\n",
       "      <td>927</td>\n",
       "      <td>1347</td>\n",
       "      <td>264</td>\n",
       "      <td>...</td>\n",
       "      <td>748</td>\n",
       "      <td>159</td>\n",
       "      <td>180</td>\n",
       "      <td>887</td>\n",
       "      <td>275</td>\n",
       "      <td>181</td>\n",
       "      <td>206</td>\n",
       "      <td>153</td>\n",
       "      <td>581</td>\n",
       "      <td>770</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 149 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    TCGA-43-5670-11A  TCGA-77-8008-11A  TCGA-18-3410-01A  \\\n",
       "feature-id                                                                 \n",
       "ENSG00000000003.13              4707              1016              2286   \n",
       "ENSG00000000005.5                  5                 1                 1   \n",
       "ENSG00000000419.11              2019              1731              2123   \n",
       "ENSG00000000457.12              1062               968              1883   \n",
       "ENSG00000000460.15               204               202              1923   \n",
       "\n",
       "                    TCGA-43-6771-11A  TCGA-66-2758-01A  TCGA-90-7767-01A  \\\n",
       "feature-id                                                                 \n",
       "ENSG00000000003.13              1238              2982              2795   \n",
       "ENSG00000000005.5                  5                 1                 0   \n",
       "ENSG00000000419.11              2121              3868              2722   \n",
       "ENSG00000000457.12               655               800              1987   \n",
       "ENSG00000000460.15               205               706              2191   \n",
       "\n",
       "                    TCGA-66-2795-01A  TCGA-77-7138-01A  TCGA-39-5019-01A  \\\n",
       "feature-id                                                                 \n",
       "ENSG00000000003.13              1516              5507              5504   \n",
       "ENSG00000000005.5                  0                 0                 2   \n",
       "ENSG00000000419.11              2130              4834              4557   \n",
       "ENSG00000000457.12               935               898              1014   \n",
       "ENSG00000000460.15               956               927              1347   \n",
       "\n",
       "                    TCGA-90-6837-11A  ...  TCGA-56-7823-01B  TCGA-77-7142-11A  \\\n",
       "feature-id                            ...                                       \n",
       "ENSG00000000003.13              2079  ...              1988              1059   \n",
       "ENSG00000000005.5                  3  ...                 1                 5   \n",
       "ENSG00000000419.11              1264  ...              1194              1449   \n",
       "ENSG00000000457.12               840  ...               449               883   \n",
       "ENSG00000000460.15               264  ...               748               159   \n",
       "\n",
       "                    TCGA-22-5483-11A  TCGA-77-8153-01A  TCGA-85-A4JC-01A  \\\n",
       "feature-id                                                                 \n",
       "ENSG00000000003.13              1297              1970              2450   \n",
       "ENSG00000000005.5                  4                 0                 0   \n",
       "ENSG00000000419.11              1382               747              1741   \n",
       "ENSG00000000457.12               614               678               305   \n",
       "ENSG00000000460.15               180               887               275   \n",
       "\n",
       "                    TCGA-39-5040-11A  TCGA-43-6143-11A  TCGA-77-7335-11A  \\\n",
       "feature-id                                                                 \n",
       "ENSG00000000003.13              1432              1500              1790   \n",
       "ENSG00000000005.5                  4                 3                 0   \n",
       "ENSG00000000419.11              3167              1454              1569   \n",
       "ENSG00000000457.12               705               917               658   \n",
       "ENSG00000000460.15               181               206               153   \n",
       "\n",
       "                    TCGA-85-7698-01A  TCGA-43-2581-01A  \n",
       "feature-id                                              \n",
       "ENSG00000000003.13              3722              2471  \n",
       "ENSG00000000005.5                  1                 0  \n",
       "ENSG00000000419.11              2477              1826  \n",
       "ENSG00000000457.12               961               470  \n",
       "ENSG00000000460.15               581               770  \n",
       "\n",
       "[5 rows x 149 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feature_table = pd.read_csv(\n",
    "    \"input/TCGA_LUSC_expression_feature_table.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=0,\n",
    ")\n",
    "print(feature_table.shape)\n",
    "feature_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(6049, 149)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-43-5670-11A</th>\n",
       "      <th>TCGA-77-8008-11A</th>\n",
       "      <th>TCGA-18-3410-01A</th>\n",
       "      <th>TCGA-43-6771-11A</th>\n",
       "      <th>TCGA-66-2758-01A</th>\n",
       "      <th>TCGA-90-7767-01A</th>\n",
       "      <th>TCGA-66-2795-01A</th>\n",
       "      <th>TCGA-77-7138-01A</th>\n",
       "      <th>TCGA-39-5019-01A</th>\n",
       "      <th>TCGA-90-6837-11A</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-56-7823-01B</th>\n",
       "      <th>TCGA-77-7142-11A</th>\n",
       "      <th>TCGA-22-5483-11A</th>\n",
       "      <th>TCGA-77-8153-01A</th>\n",
       "      <th>TCGA-85-A4JC-01A</th>\n",
       "      <th>TCGA-39-5040-11A</th>\n",
       "      <th>TCGA-43-6143-11A</th>\n",
       "      <th>TCGA-77-7335-11A</th>\n",
       "      <th>TCGA-85-7698-01A</th>\n",
       "      <th>TCGA-43-2581-01A</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feature-id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003.13</th>\n",
       "      <td>4707</td>\n",
       "      <td>1016</td>\n",
       "      <td>2286</td>\n",
       "      <td>1238</td>\n",
       "      <td>2982</td>\n",
       "      <td>2795</td>\n",
       "      <td>1516</td>\n",
       "      <td>5507</td>\n",
       "      <td>5504</td>\n",
       "      <td>2079</td>\n",
       "      <td>...</td>\n",
       "      <td>1988</td>\n",
       "      <td>1059</td>\n",
       "      <td>1297</td>\n",
       "      <td>1970</td>\n",
       "      <td>2450</td>\n",
       "      <td>1432</td>\n",
       "      <td>1500</td>\n",
       "      <td>1790</td>\n",
       "      <td>3722</td>\n",
       "      <td>2471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419.11</th>\n",
       "      <td>2019</td>\n",
       "      <td>1731</td>\n",
       "      <td>2123</td>\n",
       "      <td>2121</td>\n",
       "      <td>3868</td>\n",
       "      <td>2722</td>\n",
       "      <td>2130</td>\n",
       "      <td>4834</td>\n",
       "      <td>4557</td>\n",
       "      <td>1264</td>\n",
       "      <td>...</td>\n",
       "      <td>1194</td>\n",
       "      <td>1449</td>\n",
       "      <td>1382</td>\n",
       "      <td>747</td>\n",
       "      <td>1741</td>\n",
       "      <td>3167</td>\n",
       "      <td>1454</td>\n",
       "      <td>1569</td>\n",
       "      <td>2477</td>\n",
       "      <td>1826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000938.11</th>\n",
       "      <td>5951</td>\n",
       "      <td>6829</td>\n",
       "      <td>1067</td>\n",
       "      <td>11164</td>\n",
       "      <td>1091</td>\n",
       "      <td>492</td>\n",
       "      <td>521</td>\n",
       "      <td>335</td>\n",
       "      <td>1319</td>\n",
       "      <td>4540</td>\n",
       "      <td>...</td>\n",
       "      <td>194</td>\n",
       "      <td>3743</td>\n",
       "      <td>3615</td>\n",
       "      <td>142</td>\n",
       "      <td>1323</td>\n",
       "      <td>5410</td>\n",
       "      <td>7864</td>\n",
       "      <td>1875</td>\n",
       "      <td>805</td>\n",
       "      <td>1591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000971.14</th>\n",
       "      <td>11768</td>\n",
       "      <td>21648</td>\n",
       "      <td>1150</td>\n",
       "      <td>10141</td>\n",
       "      <td>14387</td>\n",
       "      <td>5091</td>\n",
       "      <td>2218</td>\n",
       "      <td>2186</td>\n",
       "      <td>4463</td>\n",
       "      <td>8875</td>\n",
       "      <td>...</td>\n",
       "      <td>1515</td>\n",
       "      <td>12057</td>\n",
       "      <td>5252</td>\n",
       "      <td>534</td>\n",
       "      <td>2250</td>\n",
       "      <td>4630</td>\n",
       "      <td>10018</td>\n",
       "      <td>7015</td>\n",
       "      <td>5265</td>\n",
       "      <td>8355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000001036.12</th>\n",
       "      <td>3890</td>\n",
       "      <td>3942</td>\n",
       "      <td>2134</td>\n",
       "      <td>5136</td>\n",
       "      <td>3002</td>\n",
       "      <td>1566</td>\n",
       "      <td>1757</td>\n",
       "      <td>2924</td>\n",
       "      <td>5016</td>\n",
       "      <td>4187</td>\n",
       "      <td>...</td>\n",
       "      <td>2312</td>\n",
       "      <td>3090</td>\n",
       "      <td>4425</td>\n",
       "      <td>2798</td>\n",
       "      <td>1916</td>\n",
       "      <td>3803</td>\n",
       "      <td>4518</td>\n",
       "      <td>3170</td>\n",
       "      <td>3097</td>\n",
       "      <td>3937</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 149 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    TCGA-43-5670-11A  TCGA-77-8008-11A  TCGA-18-3410-01A  \\\n",
       "feature-id                                                                 \n",
       "ENSG00000000003.13              4707              1016              2286   \n",
       "ENSG00000000419.11              2019              1731              2123   \n",
       "ENSG00000000938.11              5951              6829              1067   \n",
       "ENSG00000000971.14             11768             21648              1150   \n",
       "ENSG00000001036.12              3890              3942              2134   \n",
       "\n",
       "                    TCGA-43-6771-11A  TCGA-66-2758-01A  TCGA-90-7767-01A  \\\n",
       "feature-id                                                                 \n",
       "ENSG00000000003.13              1238              2982              2795   \n",
       "ENSG00000000419.11              2121              3868              2722   \n",
       "ENSG00000000938.11             11164              1091               492   \n",
       "ENSG00000000971.14             10141             14387              5091   \n",
       "ENSG00000001036.12              5136              3002              1566   \n",
       "\n",
       "                    TCGA-66-2795-01A  TCGA-77-7138-01A  TCGA-39-5019-01A  \\\n",
       "feature-id                                                                 \n",
       "ENSG00000000003.13              1516              5507              5504   \n",
       "ENSG00000000419.11              2130              4834              4557   \n",
       "ENSG00000000938.11               521               335              1319   \n",
       "ENSG00000000971.14              2218              2186              4463   \n",
       "ENSG00000001036.12              1757              2924              5016   \n",
       "\n",
       "                    TCGA-90-6837-11A  ...  TCGA-56-7823-01B  TCGA-77-7142-11A  \\\n",
       "feature-id                            ...                                       \n",
       "ENSG00000000003.13              2079  ...              1988              1059   \n",
       "ENSG00000000419.11              1264  ...              1194              1449   \n",
       "ENSG00000000938.11              4540  ...               194              3743   \n",
       "ENSG00000000971.14              8875  ...              1515             12057   \n",
       "ENSG00000001036.12              4187  ...              2312              3090   \n",
       "\n",
       "                    TCGA-22-5483-11A  TCGA-77-8153-01A  TCGA-85-A4JC-01A  \\\n",
       "feature-id                                                                 \n",
       "ENSG00000000003.13              1297              1970              2450   \n",
       "ENSG00000000419.11              1382               747              1741   \n",
       "ENSG00000000938.11              3615               142              1323   \n",
       "ENSG00000000971.14              5252               534              2250   \n",
       "ENSG00000001036.12              4425              2798              1916   \n",
       "\n",
       "                    TCGA-39-5040-11A  TCGA-43-6143-11A  TCGA-77-7335-11A  \\\n",
       "feature-id                                                                 \n",
       "ENSG00000000003.13              1432              1500              1790   \n",
       "ENSG00000000419.11              3167              1454              1569   \n",
       "ENSG00000000938.11              5410              7864              1875   \n",
       "ENSG00000000971.14              4630             10018              7015   \n",
       "ENSG00000001036.12              3803              4518              3170   \n",
       "\n",
       "                    TCGA-85-7698-01A  TCGA-43-2581-01A  \n",
       "feature-id                                              \n",
       "ENSG00000000003.13              3722              2471  \n",
       "ENSG00000000419.11              2477              1826  \n",
       "ENSG00000000938.11               805              1591  \n",
       "ENSG00000000971.14              5265              8355  \n",
       "ENSG00000001036.12              3097              3937  \n",
       "\n",
       "[5 rows x 149 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feature_table_filt = feature_table[\n",
    "    feature_table.sum(axis=1) > feature_table.sum(axis=1).quantile(0.9)\n",
    "]\n",
    "print(feature_table_filt.shape)\n",
    "feature_table_filt.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We also want to strip everything after the period in the Ensemble IDS. For example, `ENSG00000000003.13` should be converted to `ENSG00000000003`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-43-5670-11A</th>\n",
       "      <th>TCGA-77-8008-11A</th>\n",
       "      <th>TCGA-18-3410-01A</th>\n",
       "      <th>TCGA-43-6771-11A</th>\n",
       "      <th>TCGA-66-2758-01A</th>\n",
       "      <th>TCGA-90-7767-01A</th>\n",
       "      <th>TCGA-66-2795-01A</th>\n",
       "      <th>TCGA-77-7138-01A</th>\n",
       "      <th>TCGA-39-5019-01A</th>\n",
       "      <th>TCGA-90-6837-11A</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-56-7823-01B</th>\n",
       "      <th>TCGA-77-7142-11A</th>\n",
       "      <th>TCGA-22-5483-11A</th>\n",
       "      <th>TCGA-77-8153-01A</th>\n",
       "      <th>TCGA-85-A4JC-01A</th>\n",
       "      <th>TCGA-39-5040-11A</th>\n",
       "      <th>TCGA-43-6143-11A</th>\n",
       "      <th>TCGA-77-7335-11A</th>\n",
       "      <th>TCGA-85-7698-01A</th>\n",
       "      <th>TCGA-43-2581-01A</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feature-id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>4707</td>\n",
       "      <td>1016</td>\n",
       "      <td>2286</td>\n",
       "      <td>1238</td>\n",
       "      <td>2982</td>\n",
       "      <td>2795</td>\n",
       "      <td>1516</td>\n",
       "      <td>5507</td>\n",
       "      <td>5504</td>\n",
       "      <td>2079</td>\n",
       "      <td>...</td>\n",
       "      <td>1988</td>\n",
       "      <td>1059</td>\n",
       "      <td>1297</td>\n",
       "      <td>1970</td>\n",
       "      <td>2450</td>\n",
       "      <td>1432</td>\n",
       "      <td>1500</td>\n",
       "      <td>1790</td>\n",
       "      <td>3722</td>\n",
       "      <td>2471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>2019</td>\n",
       "      <td>1731</td>\n",
       "      <td>2123</td>\n",
       "      <td>2121</td>\n",
       "      <td>3868</td>\n",
       "      <td>2722</td>\n",
       "      <td>2130</td>\n",
       "      <td>4834</td>\n",
       "      <td>4557</td>\n",
       "      <td>1264</td>\n",
       "      <td>...</td>\n",
       "      <td>1194</td>\n",
       "      <td>1449</td>\n",
       "      <td>1382</td>\n",
       "      <td>747</td>\n",
       "      <td>1741</td>\n",
       "      <td>3167</td>\n",
       "      <td>1454</td>\n",
       "      <td>1569</td>\n",
       "      <td>2477</td>\n",
       "      <td>1826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <td>5951</td>\n",
       "      <td>6829</td>\n",
       "      <td>1067</td>\n",
       "      <td>11164</td>\n",
       "      <td>1091</td>\n",
       "      <td>492</td>\n",
       "      <td>521</td>\n",
       "      <td>335</td>\n",
       "      <td>1319</td>\n",
       "      <td>4540</td>\n",
       "      <td>...</td>\n",
       "      <td>194</td>\n",
       "      <td>3743</td>\n",
       "      <td>3615</td>\n",
       "      <td>142</td>\n",
       "      <td>1323</td>\n",
       "      <td>5410</td>\n",
       "      <td>7864</td>\n",
       "      <td>1875</td>\n",
       "      <td>805</td>\n",
       "      <td>1591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000971</th>\n",
       "      <td>11768</td>\n",
       "      <td>21648</td>\n",
       "      <td>1150</td>\n",
       "      <td>10141</td>\n",
       "      <td>14387</td>\n",
       "      <td>5091</td>\n",
       "      <td>2218</td>\n",
       "      <td>2186</td>\n",
       "      <td>4463</td>\n",
       "      <td>8875</td>\n",
       "      <td>...</td>\n",
       "      <td>1515</td>\n",
       "      <td>12057</td>\n",
       "      <td>5252</td>\n",
       "      <td>534</td>\n",
       "      <td>2250</td>\n",
       "      <td>4630</td>\n",
       "      <td>10018</td>\n",
       "      <td>7015</td>\n",
       "      <td>5265</td>\n",
       "      <td>8355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000001036</th>\n",
       "      <td>3890</td>\n",
       "      <td>3942</td>\n",
       "      <td>2134</td>\n",
       "      <td>5136</td>\n",
       "      <td>3002</td>\n",
       "      <td>1566</td>\n",
       "      <td>1757</td>\n",
       "      <td>2924</td>\n",
       "      <td>5016</td>\n",
       "      <td>4187</td>\n",
       "      <td>...</td>\n",
       "      <td>2312</td>\n",
       "      <td>3090</td>\n",
       "      <td>4425</td>\n",
       "      <td>2798</td>\n",
       "      <td>1916</td>\n",
       "      <td>3803</td>\n",
       "      <td>4518</td>\n",
       "      <td>3170</td>\n",
       "      <td>3097</td>\n",
       "      <td>3937</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 149 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 TCGA-43-5670-11A  TCGA-77-8008-11A  TCGA-18-3410-01A  \\\n",
       "feature-id                                                              \n",
       "ENSG00000000003              4707              1016              2286   \n",
       "ENSG00000000419              2019              1731              2123   \n",
       "ENSG00000000938              5951              6829              1067   \n",
       "ENSG00000000971             11768             21648              1150   \n",
       "ENSG00000001036              3890              3942              2134   \n",
       "\n",
       "                 TCGA-43-6771-11A  TCGA-66-2758-01A  TCGA-90-7767-01A  \\\n",
       "feature-id                                                              \n",
       "ENSG00000000003              1238              2982              2795   \n",
       "ENSG00000000419              2121              3868              2722   \n",
       "ENSG00000000938             11164              1091               492   \n",
       "ENSG00000000971             10141             14387              5091   \n",
       "ENSG00000001036              5136              3002              1566   \n",
       "\n",
       "                 TCGA-66-2795-01A  TCGA-77-7138-01A  TCGA-39-5019-01A  \\\n",
       "feature-id                                                              \n",
       "ENSG00000000003              1516              5507              5504   \n",
       "ENSG00000000419              2130              4834              4557   \n",
       "ENSG00000000938               521               335              1319   \n",
       "ENSG00000000971              2218              2186              4463   \n",
       "ENSG00000001036              1757              2924              5016   \n",
       "\n",
       "                 TCGA-90-6837-11A  ...  TCGA-56-7823-01B  TCGA-77-7142-11A  \\\n",
       "feature-id                         ...                                       \n",
       "ENSG00000000003              2079  ...              1988              1059   \n",
       "ENSG00000000419              1264  ...              1194              1449   \n",
       "ENSG00000000938              4540  ...               194              3743   \n",
       "ENSG00000000971              8875  ...              1515             12057   \n",
       "ENSG00000001036              4187  ...              2312              3090   \n",
       "\n",
       "                 TCGA-22-5483-11A  TCGA-77-8153-01A  TCGA-85-A4JC-01A  \\\n",
       "feature-id                                                              \n",
       "ENSG00000000003              1297              1970              2450   \n",
       "ENSG00000000419              1382               747              1741   \n",
       "ENSG00000000938              3615               142              1323   \n",
       "ENSG00000000971              5252               534              2250   \n",
       "ENSG00000001036              4425              2798              1916   \n",
       "\n",
       "                 TCGA-39-5040-11A  TCGA-43-6143-11A  TCGA-77-7335-11A  \\\n",
       "feature-id                                                              \n",
       "ENSG00000000003              1432              1500              1790   \n",
       "ENSG00000000419              3167              1454              1569   \n",
       "ENSG00000000938              5410              7864              1875   \n",
       "ENSG00000000971              4630             10018              7015   \n",
       "ENSG00000001036              3803              4518              3170   \n",
       "\n",
       "                 TCGA-85-7698-01A  TCGA-43-2581-01A  \n",
       "feature-id                                           \n",
       "ENSG00000000003              3722              2471  \n",
       "ENSG00000000419              2477              1826  \n",
       "ENSG00000000938               805              1591  \n",
       "ENSG00000000971              5265              8355  \n",
       "ENSG00000001036              3097              3937  \n",
       "\n",
       "[5 rows x 149 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feature_table_filt.index = [re.search(\"ENSG[0-9]*\", x).group() for x in feature_table_filt.index]\n",
    "feature_table_filt.index.name = \"feature-id\"\n",
    "feature_table_filt.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_table_filt.to_csv(\n",
    "    \"output/TCGA_LUSC_expression_feature_table_filt.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Running ALDEx2\n",
    "\n",
    "ALDEx2 is a tool used for differential abundance between conditions. We have provided a script to run ALDEx2 on the processed feature table. Note that this step can take several minutes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "[1] \"ALDEx2 results written to output/TCGA_LUSC_aldex_results.tsv\"\n"
     ]
    }
   ],
   "source": [
    "!Rscript run_aldex.R"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Converting feature table from tsv to biom\n",
    "\n",
    "`<Insert explanation here>`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "!biom convert \\\n",
    "    -i output/TCGA_LUSC_expression_feature_table_filt.tsv \\\n",
    "    -o output/TCGA_LUSC_expression_feature_table_filt.biom \\\n",
    "    --table-type=\"OTU table\" \\\n",
    "    --to-hdf5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Process sample metadata\n",
    "\n",
    "Qurro expects the metadata file to have the first column labelled as \"Sample ID\", so we'll edit the sample sheet to comply."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>File ID</th>\n",
       "      <th>File Name</th>\n",
       "      <th>Data Category</th>\n",
       "      <th>Data Type</th>\n",
       "      <th>Project ID</th>\n",
       "      <th>Case ID</th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Sample Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>e4c62f17-d1e8-4543-9b7e-daa2b68306e0</td>\n",
       "      <td>bc5be208-5934-40dd-81df-567599ea2a51.htseq.cou...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>TCGA-LUSC</td>\n",
       "      <td>TCGA-33-6737</td>\n",
       "      <td>TCGA-33-6737-01A</td>\n",
       "      <td>Primary Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>220a03f7-7ab6-4233-8f65-7ac5decca4b9</td>\n",
       "      <td>60040f95-8414-4956-bd8a-ec461a49207c.htseq.cou...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>TCGA-LUSC</td>\n",
       "      <td>TCGA-18-3410</td>\n",
       "      <td>TCGA-18-3410-01A</td>\n",
       "      <td>Primary Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8894c42e-ce65-4088-88e3-921ce7165261</td>\n",
       "      <td>950e2ba0-a247-4bd6-8092-f97cc4018a79.htseq.cou...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>TCGA-LUSC</td>\n",
       "      <td>TCGA-33-A4WN</td>\n",
       "      <td>TCGA-33-A4WN-01A</td>\n",
       "      <td>Primary Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>daa44ce1-1671-46b9-aa48-2f4155f0ee49</td>\n",
       "      <td>a998a5b1-397d-4497-a58c-9b9e1c7f491e.htseq.cou...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>TCGA-LUSC</td>\n",
       "      <td>TCGA-56-7579</td>\n",
       "      <td>TCGA-56-7579-01A</td>\n",
       "      <td>Primary Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ef056c34-c2b9-47dd-afbf-ed81fc16dc74</td>\n",
       "      <td>840bb854-0669-485e-9d83-c4e1e4f10626.htseq.cou...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>TCGA-LUSC</td>\n",
       "      <td>TCGA-34-5236</td>\n",
       "      <td>TCGA-34-5236-01A</td>\n",
       "      <td>Primary Tumor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                File ID  \\\n",
       "0  e4c62f17-d1e8-4543-9b7e-daa2b68306e0   \n",
       "1  220a03f7-7ab6-4233-8f65-7ac5decca4b9   \n",
       "2  8894c42e-ce65-4088-88e3-921ce7165261   \n",
       "3  daa44ce1-1671-46b9-aa48-2f4155f0ee49   \n",
       "4  ef056c34-c2b9-47dd-afbf-ed81fc16dc74   \n",
       "\n",
       "                                           File Name            Data Category  \\\n",
       "0  bc5be208-5934-40dd-81df-567599ea2a51.htseq.cou...  Transcriptome Profiling   \n",
       "1  60040f95-8414-4956-bd8a-ec461a49207c.htseq.cou...  Transcriptome Profiling   \n",
       "2  950e2ba0-a247-4bd6-8092-f97cc4018a79.htseq.cou...  Transcriptome Profiling   \n",
       "3  a998a5b1-397d-4497-a58c-9b9e1c7f491e.htseq.cou...  Transcriptome Profiling   \n",
       "4  840bb854-0669-485e-9d83-c4e1e4f10626.htseq.cou...  Transcriptome Profiling   \n",
       "\n",
       "                        Data Type Project ID       Case ID         Sample ID  \\\n",
       "0  Gene Expression Quantification  TCGA-LUSC  TCGA-33-6737  TCGA-33-6737-01A   \n",
       "1  Gene Expression Quantification  TCGA-LUSC  TCGA-18-3410  TCGA-18-3410-01A   \n",
       "2  Gene Expression Quantification  TCGA-LUSC  TCGA-33-A4WN  TCGA-33-A4WN-01A   \n",
       "3  Gene Expression Quantification  TCGA-LUSC  TCGA-56-7579  TCGA-56-7579-01A   \n",
       "4  Gene Expression Quantification  TCGA-LUSC  TCGA-34-5236  TCGA-34-5236-01A   \n",
       "\n",
       "     Sample Type  \n",
       "0  Primary Tumor  \n",
       "1  Primary Tumor  \n",
       "2  Primary Tumor  \n",
       "3  Primary Tumor  \n",
       "4  Primary Tumor  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_sheet = pd.read_csv(\"input/gdc_sample_sheet.2020-02-13.tsv\", sep=\"\\t\")\n",
    "sample_sheet.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>File ID</th>\n",
       "      <th>File Name</th>\n",
       "      <th>Data Category</th>\n",
       "      <th>Data Type</th>\n",
       "      <th>Project ID</th>\n",
       "      <th>Case ID</th>\n",
       "      <th>Sample Type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-33-6737-01A</th>\n",
       "      <td>e4c62f17-d1e8-4543-9b7e-daa2b68306e0</td>\n",
       "      <td>bc5be208-5934-40dd-81df-567599ea2a51.htseq.cou...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>TCGA-LUSC</td>\n",
       "      <td>TCGA-33-6737</td>\n",
       "      <td>Primary Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-18-3410-01A</th>\n",
       "      <td>220a03f7-7ab6-4233-8f65-7ac5decca4b9</td>\n",
       "      <td>60040f95-8414-4956-bd8a-ec461a49207c.htseq.cou...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>TCGA-LUSC</td>\n",
       "      <td>TCGA-18-3410</td>\n",
       "      <td>Primary Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-33-A4WN-01A</th>\n",
       "      <td>8894c42e-ce65-4088-88e3-921ce7165261</td>\n",
       "      <td>950e2ba0-a247-4bd6-8092-f97cc4018a79.htseq.cou...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>TCGA-LUSC</td>\n",
       "      <td>TCGA-33-A4WN</td>\n",
       "      <td>Primary Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-56-7579-01A</th>\n",
       "      <td>daa44ce1-1671-46b9-aa48-2f4155f0ee49</td>\n",
       "      <td>a998a5b1-397d-4497-a58c-9b9e1c7f491e.htseq.cou...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>TCGA-LUSC</td>\n",
       "      <td>TCGA-56-7579</td>\n",
       "      <td>Primary Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-34-5236-01A</th>\n",
       "      <td>ef056c34-c2b9-47dd-afbf-ed81fc16dc74</td>\n",
       "      <td>840bb854-0669-485e-9d83-c4e1e4f10626.htseq.cou...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>TCGA-LUSC</td>\n",
       "      <td>TCGA-34-5236</td>\n",
       "      <td>Primary Tumor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               File ID  \\\n",
       "Sample ID                                                \n",
       "TCGA-33-6737-01A  e4c62f17-d1e8-4543-9b7e-daa2b68306e0   \n",
       "TCGA-18-3410-01A  220a03f7-7ab6-4233-8f65-7ac5decca4b9   \n",
       "TCGA-33-A4WN-01A  8894c42e-ce65-4088-88e3-921ce7165261   \n",
       "TCGA-56-7579-01A  daa44ce1-1671-46b9-aa48-2f4155f0ee49   \n",
       "TCGA-34-5236-01A  ef056c34-c2b9-47dd-afbf-ed81fc16dc74   \n",
       "\n",
       "                                                          File Name  \\\n",
       "Sample ID                                                             \n",
       "TCGA-33-6737-01A  bc5be208-5934-40dd-81df-567599ea2a51.htseq.cou...   \n",
       "TCGA-18-3410-01A  60040f95-8414-4956-bd8a-ec461a49207c.htseq.cou...   \n",
       "TCGA-33-A4WN-01A  950e2ba0-a247-4bd6-8092-f97cc4018a79.htseq.cou...   \n",
       "TCGA-56-7579-01A  a998a5b1-397d-4497-a58c-9b9e1c7f491e.htseq.cou...   \n",
       "TCGA-34-5236-01A  840bb854-0669-485e-9d83-c4e1e4f10626.htseq.cou...   \n",
       "\n",
       "                            Data Category                       Data Type  \\\n",
       "Sample ID                                                                   \n",
       "TCGA-33-6737-01A  Transcriptome Profiling  Gene Expression Quantification   \n",
       "TCGA-18-3410-01A  Transcriptome Profiling  Gene Expression Quantification   \n",
       "TCGA-33-A4WN-01A  Transcriptome Profiling  Gene Expression Quantification   \n",
       "TCGA-56-7579-01A  Transcriptome Profiling  Gene Expression Quantification   \n",
       "TCGA-34-5236-01A  Transcriptome Profiling  Gene Expression Quantification   \n",
       "\n",
       "                 Project ID       Case ID    Sample Type  \n",
       "Sample ID                                                 \n",
       "TCGA-33-6737-01A  TCGA-LUSC  TCGA-33-6737  Primary Tumor  \n",
       "TCGA-18-3410-01A  TCGA-LUSC  TCGA-18-3410  Primary Tumor  \n",
       "TCGA-33-A4WN-01A  TCGA-LUSC  TCGA-33-A4WN  Primary Tumor  \n",
       "TCGA-56-7579-01A  TCGA-LUSC  TCGA-56-7579  Primary Tumor  \n",
       "TCGA-34-5236-01A  TCGA-LUSC  TCGA-34-5236  Primary Tumor  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_sheet_new = sample_sheet.set_index(\"Sample ID\", drop=True)\n",
    "sample_sheet_new.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_sheet_new.to_csv(\"output/sample_metadata.tsv\", sep=\"\\t\", index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Mapping Ensembl gene identifiers to HGNC\n",
    "\n",
    "We might want to know which Ensembl features map to which HUGO features. First, we'll download the `GRCh38` gtf file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2020-02-14 16:07:11--  ftp://ftp.ensembl.org/pub/release-99/gtf/homo_sapiens/Homo_sapiens.GRCh38.99.gtf.gz\n",
      "           => ‘output/Homo_sapiens.GRCh38.99.gtf.gz’\n",
      "Resolving ftp.ensembl.org (ftp.ensembl.org)... 193.62.193.8\n",
      "Connecting to ftp.ensembl.org (ftp.ensembl.org)|193.62.193.8|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /pub/release-99/gtf/homo_sapiens ... done.\n",
      "==> SIZE Homo_sapiens.GRCh38.99.gtf.gz ... 46905912\n",
      "==> PASV ... done.    ==> RETR Homo_sapiens.GRCh38.99.gtf.gz ... done.\n",
      "Length: 46905912 (45M) (unauthoritative)\n",
      "\n",
      "Homo_sapiens.GRCh38 100%[===================>]  44.73M  2.62MB/s    in 18s     \n",
      "\n",
      "2020-02-14 16:07:31 (2.52 MB/s) - ‘output/Homo_sapiens.GRCh38.99.gtf.gz’ saved [46905912]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget -P output/ ftp://ftp.ensembl.org/pub/release-99/gtf/homo_sapiens/Homo_sapiens.GRCh38.99.gtf.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we'll create a feature metadata file mapping the Ensembl ID to the gene name."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "echo -e \"feature-id\\tGene Name\" > output/gene_matching.tsv\n",
    "\n",
    "zgrep \"havana\\tgene\" output/Homo_sapiens.GRCh38.99.gtf.gz | \\\n",
    "awk -F \"\\t\" '{print $9}' | \\\n",
    "awk -F \";\" -v OFS=\"\\t\" '{print $1, $3}' | \\\n",
    "sed -e \"s/gene_[a-z]* //g\" | \\\n",
    "tr -d \\\" >> output/gene_matching.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Running Qurro\n",
    "\n",
    "Finally, we'll run Qurro."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Successfully generated a visualization in the folder qurro.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/miniconda3/envs/qurro_revisions/lib/python3.6/site-packages/qurro/_df_utils.py:126: FutureWarning: SparseDataFrame is deprecated and will be removed in a future version.\n",
      "Use a regular DataFrame whose columns are SparseArrays instead.\n",
      "\n",
      "See http://pandas.pydata.org/pandas-docs/stable/user_guide/sparse.html#migrating for more.\n",
      "\n",
      "  table_sdf = pd.SparseDataFrame(table.matrix_data, default_fill_value=0.0)\n",
      "/miniconda3/envs/qurro_revisions/lib/python3.6/site-packages/pandas/core/sparse/frame.py:257: FutureWarning: SparseSeries is deprecated and will be removed in a future version.\n",
      "Use a Series with sparse values instead.\n",
      "\n",
      "    >>> series = pd.Series(pd.SparseArray(...))\n",
      "\n",
      "See http://pandas.pydata.org/pandas-docs/stable/user_guide/sparse.html#migrating for more.\n",
      "\n",
      "  sparse_index=BlockIndex(N, blocs, blens),\n",
      "/miniconda3/envs/qurro_revisions/lib/python3.6/site-packages/pandas/core/frame.py:3471: FutureWarning: SparseSeries is deprecated and will be removed in a future version.\n",
      "Use a Series with sparse values instead.\n",
      "\n",
      "    >>> series = pd.Series(pd.SparseArray(...))\n",
      "\n",
      "See http://pandas.pydata.org/pandas-docs/stable/user_guide/sparse.html#migrating for more.\n",
      "\n",
      "  return klass(values, index=self.index, name=items, fastpath=True)\n",
      "/miniconda3/envs/qurro_revisions/lib/python3.6/site-packages/pandas/core/ops/__init__.py:1641: FutureWarning: SparseSeries is deprecated and will be removed in a future version.\n",
      "Use a Series with sparse values instead.\n",
      "\n",
      "    >>> series = pd.Series(pd.SparseArray(...))\n",
      "\n",
      "See http://pandas.pydata.org/pandas-docs/stable/user_guide/sparse.html#migrating for more.\n",
      "\n",
      "  return self._constructor(new_values, index=self.index, name=self.name)\n",
      "/miniconda3/envs/qurro_revisions/lib/python3.6/site-packages/pandas/core/sparse/frame.py:339: FutureWarning: SparseDataFrame is deprecated and will be removed in a future version.\n",
      "Use a regular DataFrame whose columns are SparseArrays instead.\n",
      "\n",
      "See http://pandas.pydata.org/pandas-docs/stable/user_guide/sparse.html#migrating for more.\n",
      "\n",
      "  default_fill_value=self.default_fill_value,\n",
      "/miniconda3/envs/qurro_revisions/lib/python3.6/site-packages/pandas/core/generic.py:6289: FutureWarning: SparseDataFrame is deprecated and will be removed in a future version.\n",
      "Use a regular DataFrame whose columns are SparseArrays instead.\n",
      "\n",
      "See http://pandas.pydata.org/pandas-docs/stable/user_guide/sparse.html#migrating for more.\n",
      "\n",
      "  return self._constructor(new_data).__finalize__(self)\n",
      "/miniconda3/envs/qurro_revisions/lib/python3.6/site-packages/pandas/core/generic.py:5884: FutureWarning: SparseSeries is deprecated and will be removed in a future version.\n",
      "Use a Series with sparse values instead.\n",
      "\n",
      "    >>> series = pd.Series(pd.SparseArray(...))\n",
      "\n",
      "See http://pandas.pydata.org/pandas-docs/stable/user_guide/sparse.html#migrating for more.\n",
      "\n",
      "  return self._constructor(new_data).__finalize__(self)\n",
      "/miniconda3/envs/qurro_revisions/lib/python3.6/site-packages/pandas/core/sparse/frame.py:785: FutureWarning: SparseDataFrame is deprecated and will be removed in a future version.\n",
      "Use a regular DataFrame whose columns are SparseArrays instead.\n",
      "\n",
      "See http://pandas.pydata.org/pandas-docs/stable/user_guide/sparse.html#migrating for more.\n",
      "\n",
      "  return self._constructor(new_arrays, index=index, columns=columns).__finalize__(\n",
      "/miniconda3/envs/qurro_revisions/lib/python3.6/site-packages/pandas/core/generic.py:3606: FutureWarning: SparseDataFrame is deprecated and will be removed in a future version.\n",
      "Use a regular DataFrame whose columns are SparseArrays instead.\n",
      "\n",
      "See http://pandas.pydata.org/pandas-docs/stable/user_guide/sparse.html#migrating for more.\n",
      "\n",
      "  result = self._constructor(new_data).__finalize__(self)\n",
      "/miniconda3/envs/qurro_revisions/lib/python3.6/site-packages/pandas/core/generic.py:1999: FutureWarning: SparseDataFrame is deprecated and will be removed in a future version.\n",
      "Use a regular DataFrame whose columns are SparseArrays instead.\n",
      "\n",
      "See http://pandas.pydata.org/pandas-docs/stable/user_guide/sparse.html#migrating for more.\n",
      "\n",
      "  return self._constructor(result, **d).__finalize__(self)\n",
      "/miniconda3/envs/qurro_revisions/lib/python3.6/site-packages/pandas/core/sparse/frame.py:745: FutureWarning: SparseDataFrame is deprecated and will be removed in a future version.\n",
      "Use a regular DataFrame whose columns are SparseArrays instead.\n",
      "\n",
      "See http://pandas.pydata.org/pandas-docs/stable/user_guide/sparse.html#migrating for more.\n",
      "\n",
      "  default_fill_value=self._default_fill_value,\n",
      "/miniconda3/envs/qurro_revisions/lib/python3.6/site-packages/pandas/core/generic.py:9126: FutureWarning: SparseDataFrame is deprecated and will be removed in a future version.\n",
      "Use a regular DataFrame whose columns are SparseArrays instead.\n",
      "\n",
      "See http://pandas.pydata.org/pandas-docs/stable/user_guide/sparse.html#migrating for more.\n",
      "\n",
      "  return self._constructor(new_data).__finalize__(self)\n",
      "/miniconda3/envs/qurro_revisions/lib/python3.6/site-packages/pandas/core/sparse/frame.py:854: FutureWarning: SparseDataFrame is deprecated and will be removed in a future version.\n",
      "Use a regular DataFrame whose columns are SparseArrays instead.\n",
      "\n",
      "See http://pandas.pydata.org/pandas-docs/stable/user_guide/sparse.html#migrating for more.\n",
      "\n",
      "  default_kind=self._default_kind,\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "qurro \\\n",
    "    -r output/TCGA_LUSC_aldex_results.tsv \\\n",
    "    -t output/TCGA_LUSC_expression_feature_table_filt.biom \\\n",
    "    -sm output/sample_metadata.tsv \\\n",
    "    -fm output/gene_matching.tsv \\\n",
    "    -o qurro"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
